homeassistant/components/ovo_energy/__init__.py

```diff
@@ -0,0 +1,151 @@
+    """Set up OVO Energy from a config entry."""
+
+    client = OVOEnergy()
+
+    try:
+        await client.authenticate(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
+    except aiohttp.ClientError as exception:
+        _LOGGER.warning(exception)
+        raise ConfigEntryNotReady from exception
+
+    async def async_update_data() -> OVODailyUsage:
+        """Fetch data from OVO Energy."""
+        now = datetime.utcnow()
+        async with async_timeout.timeout(10):
+            return await client.get_daily_usage(now.strftime("%Y-%m"))
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="sensor",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=300),
+    )
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_CLIENT: client,
+        DATA_COORDINATOR: coordinator,
+    }
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    # Setup components
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, "sensor")
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigType) -> bool:
+    """Unload OVO Energy config entry."""
+    # Unload sensors
+    await hass.config_entries.async_forward_entry_unload(entry, "sensor")
+
+    del hass.data[DOMAIN][entry.entry_id]
+
+    return True
+
+
+class OVOEnergyEntity(Entity):
+    """Defines a base OVO Energy entity."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        client: OVOEnergy,
+        key: str,
+        name: str,
+        icon: str,
+    ) -> None:
+        """Initialize the OVO Energy entity."""
+        self._coordinator = coordinator
+        self._client = client
+        self._key = key
+        self._name = name
+        self._icon = icon
+        self._available = True
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return self._key
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def icon(self) -> str:
+        """Return the mdi icon of the entity."""
+        return self._icon
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._coordinator.last_update_success and self._available
+
+    @property
+    def should_poll(self):
+        """No need to poll. Coordinator notifies entity of updates."""
+        return False
+
+    async def async_update(self) -> None:
+        """Update OVO Energy entity."""
+        if await self._ovo_energy_update():
+            self._available = True
+        else:
+            self._available = False
```

Comment:
Given the above example, the coordinator can be used to ask for an update. After the coordinator has updated its data, it will notify the entities.

```suggestion
    async def async_update(self) -> None:
        """Update OVO Energy entity."""
        await self.coordinator.async_request_refresh()
```